### ISSUE TITLE
Ignoring class methods with underscore prefix cause issue with some properties

### ISSUE THREAD

**Issue by touilleMan** [godotengine/godot#64429]
### Godot version

4.0-rc13

### System information

n/a

### Issue description

Currently the `extension_api.json` generation code consider in class methods/properties starting with an underscore as hidden and ignore it:
 https://github.com/godotengine/godot/blob/eda6800f5dc01981f3626580d91a619b496acb5f/core/extension/extension_api_dump.cpp#L748-L749

https://github.com/godotengine/godot/blob/eda6800f5dc01981f3626580d91a619b496acb5f/core/extension/extension_api_dump.cpp#L846-L848

However _ as a prefix is not always used to indicate a private method, for instance considering the `anchor_*` properties in `Control`:
https://github.com/godotengine/godot/blob/eda6800f5dc01981f3626580d91a619b496acb5f/scene/gui/control.cpp#L3355-L3358

Those properties use `_set_anchor` as setter, this is because `Control` also expose a `set_anchor` method (this one behind higher level designed to be directly used)
https://github.com/godotengine/godot/blob/eda6800f5dc01981f3626580d91a619b496acb5f/scene/gui/control.cpp#L3189

So my guess is we should stop considering the underscore prefix as a private marker (and I'm saying this as a Python lover :smile: ) and just dump all methods/properties.

As a matter of fact I even wonder if there is such thing  as class method declared in ClassDB that should be private (the whole point of GDExtension being to provide to external tools the same access GDScript have)

### Steps to reproduce

n/a

**Comment by Splizard** [comment]
The `_` prefix for methods indicates that the method is virtual and can be overridden by a custom implementation.

**Comment by touilleMan** [comment]
@Splizard Thanks for the feedback

I have the felling this "underscore for virtual method" is very error prone given the is no unittests preventing us to do a bad naming. So we should not rely in that to filter out methods.

On top of that the comment in `extension_api_dump.cpp` is talking about *hidden* methods which is a wrong if it stands for *virtual*

**Comment by akien-mga** [comment]
A leading underscore is typically used for one of two meanings:
- A virtual method, which is flagged as `virtual` and this information should be available to GDExtension, it doesn't need to be guessed based on naming.
- A method that needs to be exposed but which we don't want visible in the API. That's typically for methods which need to be called by string (e.g. for UndoRedo). This also used to include all signal callbacks in `3.x` but now most of those are using `callable_mp` and no longer need to be bound. Those are the methods meant to be "hidden" because while they're exposed and public, they're not meant to be in theory.

**Comment by touilleMan** [comment]
> A method that needs to be exposed but which we don't want visible in the API.

Don't we need a dedicated flag for this kind of job ? (just like we have the virtual flag as you say)

Without a specific `private` flag, that means the test for private is:.
```cpp
is_private = F.name.begins_with("_") && !(F.flags & METHOD_FLAG_VIRTUAL)
```

This seems error prone and hard to understand :/

**Comment by akien-mga** [comment]
We could maybe add one, yeah.

This kind of check is indeed done e.g. in `editor/doc_tools.cpp` to hide those methods from the class reference:
```
if (E.name.is_empty() || (E.name[0] == '_' && !(E.flags & METHOD_FLAG_VIRTUAL))) {
	continue; //hidden, don't count
}
```

Maybe `is_internal` or something like this. It's not `private` in the common sense because even though we hide those where we can, you _can_ call these methods.

**Comment by reduz** [comment]
The general rule is that, anything with starts with '_' is either private or protected. Virtual functions are protected, neither can be accessed from outside the class.

If you are creating a binder, the logic should be:
* If your language supports properties **do not** expose the setters and getters to it, only the properties.
* If your language _does not_ support properties, **ignore them** and expose any non private method.
* Virtual methods are protected, hence always exposed.

**Comment by AwesomeJuniorMiss** [comment]
So that's why we can't call super._process(delta) from GDScript that extends GDExtension class? But in many other engines, like Unreal, it's commonly used and can be very useful.

**Comment by dsnopek** [comment]
> So that's why we can't call super._process(delta) from GDScript that extends GDExtension class? But in many other engines, like Unreal, it's commonly used and can be very useful.

No, I think that's unrelated. The problem is that we don't really have virtual methods defined in GDExtension yet, and `super` in GDScript can only really refer to methods on parent GDScript classes - I don't think it works with native classes either. A bunch of additional work around virtual methods would need to be done to allow `super` in GDScript to work across languages.
